eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h

```diff
@@ -18,103 +54,57 @@ namespace vespalib::tensor {
+    using Mapping = AddressContext::Mapping;
 
 private:
     using Address = DenseTensorCellsIterator::Address;
     using CellsRef = vespalib::ConstArrayRef<double>;
     using size_type = eval::ValueType::Dimension::size_type;
 
-    class AddressContext {
-    public:
-        AddressContext(const eval::ValueType &type);
-        ~AddressContext();
-        size_type dimSize(uint32_t dim) const { return _type.dimensions()[dim].size; }
-        size_type wholeDimStep(uint32_t dim) const { return _accumulatedSize[dim] * dimSize(dim); }
-        size_t index() const {
-            size_t cellIdx(0);
-            for (uint32_t i(0); i < _address.size(); i++) {
-                cellIdx += _address[i]*_accumulatedSize[i];
-            }
-            return cellIdx;
-        }
-        void update(const Address & addr, const Mapping & mapping) {
-            for (const auto & m : mapping) {
-                _address[m.first] = addr[m.second];
-            }
-        }
-
-        const eval::ValueType &_type;
-        std::vector<size_t>    _accumulatedSize;
-        Address                _address;
-
-    };
-    class CellAddressContext : public AddressContext {
-    public:
-        CellAddressContext(const eval::ValueType &type, CellsRef cells) : AddressContext(type), _cells(cells) { }
-        double cell() const { return cell(index()); }
-        double cell(size_t cellIdx) const { return _cells[cellIdx]; }
-    private:
-        CellsRef               _cells;
-    };
-
-
-    CellAddressContext     _rightAddress;
     AddressContext         _combinedAddress;
 
     Mapping                _left;
     Mapping                _commonRight;
     Mapping                _right;
 
 public:
-    DenseTensorAddressCombiner(const eval::ValueType &combined, const eval::ValueType &lhs,
-                               const eval::ValueType &rhs, CellsRef rhsCells);
+    DenseTensorAddressCombiner(const eval::ValueType &combined, const eval::ValueType &lhs, const eval::ValueType &rhs);
     ~DenseTensorAddressCombiner();
     void updateLeftAndCommon(const Address & addr) { _combinedAddress.update(addr, _left); }
 
     bool hasAnyRightOnlyDimensions() const { return ! _right.empty(); }
 
-    const Address &address() const { return _combinedAddress._address; }
-
-    bool updateCommonRight() {
-        for (const auto & m : _commonRight) {
-            if (_combinedAddress._address[m.first] >= _rightAddress.dimSize(m.second)) {
-                return false;
-            }
-            _rightAddress._address[m.second] = _combinedAddress._address[m.first];
-        }
-        return true;
-    }
-    double rightCell() { return _rightAddress.cell(); }
+    const Address & address() const { return _combinedAddress._address; }
+    const Mapping & commonRight() const { return _commonRight; }
 
     template <typename Func>
-    void for_each(Func && func) {
+    void for_each(const AddressContext & rightAddress, const CellsRef & rhsCells, Func && func) {
```

Comment:
I actually think the rhs address should stay inside the addresscombiner (only the rhs cell references should be on the outside)